waspc/packages/wasp-config/src/mappers.ts

```diff
@@ -0,0 +1,443 @@
+/** This module maps the user-facing API to the internal representation of the app (AppSpec Decl).
+ */
+import * as AppSpec from './appSpec.js'
+import * as User from './userApi.js'
+
+export function mapUserSpecToDecls(
+  spec: User.UserSpec,
+  entityNames: string[]
+): AppSpec.Decl[] {
+  const {
+    app,
+    actions,
+    apis,
+    apiNamespaces,
+    auth,
+    client,
+    db,
+    emailSender,
+    jobs,
+    pages,
+    queries,
+    routes,
+    server,
+    websocket,
+  } = spec
+
+  const pageNames = Array.from(pages.keys())
+  const routeNames = Array.from(routes.keys())
+  const parseEntityRef = makeRefParser('Entity', entityNames)
+  const parsePageRef = makeRefParser('Page', pageNames)
+  const parseRouteRef = makeRefParser('Route', routeNames)
+
+  // TODO: Try to build the entire object at once
+  const decls: AppSpec.Decl[] = []
+
+  // TODO: Find a way to make sure you've covered everything in compile time
+  for (const [pageName, pageConfig] of pages.entries()) {
+    decls.push({
+      declType: 'Page',
+      declName: pageName,
+      declValue: mapPage(pageConfig),
+    })
+  }
+
+  for (const [routeName, routeConfig] of routes.entries()) {
+    decls.push({
+      declType: 'Route',
+      declName: routeName,
+      declValue: mapRoute(routeConfig, parsePageRef),
+    })
+  }
+
+  decls.push({
+    declType: 'App',
+    declName: app.name,
+    declValue: mapApp(
+      app.config,
+      parseEntityRef,
+      parseRouteRef,
+      auth,
+      server,
+      client,
+      db,
+      emailSender,
+      websocket
+    ),
+  })
+
+  for (const [actionName, actionConfig] of actions.entries()) {
+    decls.push({
+      declType: 'Action',
+      declName: actionName,
+      declValue: mapOperationConfig(actionConfig, parseEntityRef),
+    })
+  }
```

Comment:
These for loops are also repetitive, and I'll probably rewrite them as a generic map, and add a compile-time check making sure I've covered all the options.

But let's release it first and then deal with that.

Reply:
Sure, I in general just skimmed this file, let's release and then we can work on making it perfect. I am sure we can work on naming, code itself, readability (splitting it up a bit), ... .

Reply:
Took care of this in https://github.com/wasp-lang/wasp/pull/2393